File: example/flux_example_gpu/flux_example_gpu/image_sub_gpu.py
#!/usr/bin/env python3
"""CUDA subscriber for the same host publisher: no copy, no staging, no handle.

ros2 run flux_example_gpu image_sub_gpu.py

The publisher need never mention the GPU. On an integrated GPU the slot mapping already is device
memory, so device="cuda" is a local statement about who reads the bytes -- nothing on the wire
changes, and a host subscriber keeps running unmodified.

Frames arrive as flux.Frame instead of numpy. The `with` domain is required: leaving it waits on
the stream before releasing the borrow, and that wait must happen where the caller put it rather
than whenever Python collects the object.
"""

import flux.ros
import rclpy
from rclpy.executors import ExternalShutdownException
from rclpy.node import Node


class GpuImageSubscriber(Node):

    TOPIC = "image_gpu"

    def __init__(self):
        super().__init__("flux_gpu_image_sub")
        self.seen = 0
        self.device_ptr = 0
        self.shape = ()
        self.subscription = flux.ros.Subscription(
            self, self.TOPIC, callback=self.on_frame, fingerprint=flux.NO_SCHEMA, device="cuda"
        )
        self.timer = self.create_timer(1.0, self.report)

    def on_frame(self, frame):
        with frame as view:
            self.device_ptr = view.__cuda_array_interface__["data"][0]
            self.shape = view.shape
            self.seen += 1

    def report(self):
        self.get_logger().info(
            f"seen {self.seen}  {self.shape}  device_ptr 0x{self.device_ptr:x}  "
            f"release max {self.subscription.fence_wait.release_max_ns / 1e3:.0f} us  "
            f"fence_failed {self.subscription.fence_failed}"
        )


def main(args=None):
    rclpy.init(args=args)
    try:
        node = GpuImageSubscriber()
    except ValueError as exc:
        rclpy.logging.get_logger("flux_gpu_image_sub").error(str(exc))
        rclpy.try_shutdown()
        return
    executor = flux.ros.Executor()
    executor.add_flux(node.subscription)
    executor.add_ros_node(node)
    try:
        executor.spin()
    except (KeyboardInterrupt, ExternalShutdownException):
        pass
    finally:
        executor.stop()
        executor.close()
        node.destroy_node()
        rclpy.try_shutdown()


if __name__ == "__main__":
    main()

File: example/flux_example_gpu/src/image_sub_gpu.cpp
#include "flux/ros/executor.hpp"
#include "flux/ros/subscription.hpp"

#include <rclcpp/rclcpp.hpp>

#include <chrono>
#include <cstdint>
#include <memory>
#include <stdexcept>

// The release seam is a stream synchronization: dropping the view waits for the consuming kernel
// before the publisher may reuse the slot. fence_wait() is what that costs, and a nonzero
// fence_failed() is a fault, not a rate -- each one leaks a slot that is never reused.
class GpuImageSubscriber : public rclcpp::Node
{
public:
  GpuImageSubscriber()
  : Node("flux_gpu_image_sub"),
    sub_(
      *this, kTopic, flux::kNoSchema, [this](const flux::FrameView & f) { on_frame(f); },
      flux::QoS{}, flux::Device::Cuda)
  {
    timer_ = create_wall_timer(std::chrono::seconds(1), [this]() { report(); });
  }

  flux::ros::Subscription & subscription() { return sub_; }

private:
  static constexpr const char * kTopic = "image_gpu";

  void on_frame(const flux::FrameView & f)
  {
    device_ptr_ = f.device_ptr();
    ++seen_;
  }

  void report()
  {
    const flux::Channel::FenceWait fw = sub_.fence_wait();
    RCLCPP_INFO(
      get_logger(), "seen %lu  device_ptr %p  release max %lu us  fence_failed %lu", seen_,
      device_ptr_, fw.release_max_ns / 1000, sub_.fence_failed());
  }

  flux::ros::Subscription sub_;
  rclcpp::TimerBase::SharedPtr timer_;
  std::uint64_t seen_ = 0;
  const void * device_ptr_ = nullptr;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  try {
    auto node = std::make_shared<GpuImageSubscriber>();

    flux::ros::Executor ex;
    ex.add(node->subscription());
    ex.add_ros_node(node);

    rclcpp::on_shutdown([&ex]() { ex.stop(); });
    ex.spin();
  } catch (const std::invalid_argument & e) {
    RCLCPP_ERROR(rclcpp::get_logger("flux_gpu_image_sub"), "%s", e.what());
    rclcpp::shutdown();
    return 1;
  }
  rclcpp::shutdown();
  return 0;
}
